nativeunwind/elfunwindinfo/elfehframetable.go

```diff
@@ -0,0 +1,142 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package elfunwindinfo // import "go.opentelemetry.io/ebpf-profiler/nativeunwind/elfunwindinfo"
+
+import (
+	"debug/elf"
+	"errors"
+	"fmt"
+	"sort"
+	"unsafe"
+
+	lru "github.com/elastic/go-freelru"
+	"go.opentelemetry.io/ebpf-profiler/libpf"
+	"go.opentelemetry.io/ebpf-profiler/libpf/pfelf"
+)
+
+type FDE struct {
+	PCBegin uintptr
+	PCRange uintptr
+}
+
+type EhFrameTable struct {
+	r             reader
+	hdr           *ehFrameHdr
+	fdeCount      uintptr
+	tableStartPos uintptr
+	ehFrameSec    *elfRegion
+	efm           elf.Machine
+	cieCache      *lru.LRU[uint64, *cieInfo]
+}
+
+func NewEhFrameTable(ef *pfelf.File) (EhFrameTable, error) {
+	ehFrameHdrSec, ehFrameSec, err := findEhSections(ef)
+	if err != nil {
+		return EhFrameTable{}, fmt.Errorf("failed to get EH sections: %w", err)
+	}
+	if ehFrameSec == nil {
+		return EhFrameTable{}, errors.New(".eh_frame not found")
+	}
+	if ehFrameHdrSec == nil {
+		return EhFrameTable{}, errors.New(".eh_frame_hdr not found")
+	}
+	return newEhFrameTableFromSections(ehFrameHdrSec, ehFrameSec, ef.Machine)
+}
+
+// LookupFDE performs a binary search in .eh_frame_hdr for an FDE covering the given addr.
+func (e *EhFrameTable) LookupFDE(addr libpf.Address) (FDE, error) {
+	idx := sort.Search(e.count(), func(idx int) bool {
+		e.position(idx)
+		ipStart, _, _ := e.parseHdrEntry() // ignoring error, check bounds later
+		return ipStart > uintptr(addr)
+	})
+	if idx <= 0 {
+		return FDE{}, errors.New("FDE not found")
+	}
+	e.position(idx - 1)
+	ipStart, fr, entryErr := e.parseHdrEntry()
+	if entryErr != nil {
+		return FDE{}, entryErr
+	}
+	_, fde, _, err := parsesFDEHeader(&fr, e.efm, ipStart, e.cieCache)
+	if err != nil {
+		return FDE{}, err
+	}
+	if uintptr(addr) < fde.ipStart || uintptr(addr) >= fde.ipStart+fde.ipLen {
+		return FDE{}, errors.New("FDE not found")
+	}
+
+	return FDE{
+		PCBegin: fde.ipStart,
+		PCRange: fde.ipLen,
+	}, nil
+}
+
+func newEhFrameTableFromSections(ehFrameHdrSec *elfRegion,
+	ehFrameSec *elfRegion, efm elf.Machine,
+) (hp EhFrameTable, err error) {
```

Comment:
Similar to the comment on `NewEhFrameTable()`. In error cases `nil` should be returned instead of a half initialized `EhFrameTable` struct.
```suggestion
) (hp *EhFrameTable, err error) {
```
Returning `*EhFrameTable` should also help the Go GC to properly handle the element and clean it up.